RuntimeException when using `calculateResolvedType` for method call to `Stream.collect`

I am using JavaParser and JavaSymbolResolver to check Java code for some special kinds of method calls.

When processing the method call `stream.collect(...)` in the following code:
```
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ClassFoo {

    Map<String, Long> bar(Stream<String> stream) {
        return stream.collect(Collectors.groupingBy(String::new, Collectors.counting()));
    }
}
```
the JavaSymbolSolver runs into a `RuntimeException` with the message
```
Error calculating the type of parameter Collectors.groupingBy(String::new, Collectors.counting()) of method call stream.collect(Collectors.groupingBy(String::new, Collectors.counting()))
```

# Steps to reproduce

I simplified the problematic part of my analyser into a JUnit test that demonstrates the problem:
```
import com.github.javaparser.StaticJavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.expr.MethodCallExpr;
import com.github.javaparser.resolution.types.ResolvedType;
import com.github.javaparser.symbolsolver.JavaSymbolSolver;
import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class ParameterTypeNotCalculatedTest {
    private static final String FILE_TEXT = "" +
            "import java.util.Map;\n" +
            "import java.util.stream.Collectors;\n" +
            "import java.util.stream.Stream;\n" +
            "\n" +
            "public class ClassFoo {\n" +
            "\n" +
            "    Map<String, Long> bar(Stream<String> stream) {\n" +
            "        return stream.collect(Collectors.groupingBy(String::new, Collectors.counting()));\n" +
            "    }\n" +
            "}\n";

    @Test
    void calculateResolvedTypeForStreamCollectCall() {
        TypeSolver typeSolver = new ReflectionTypeSolver();
        JavaSymbolSolver symbolSolver = new JavaSymbolSolver(typeSolver);
        StaticJavaParser
                .getConfiguration()
                .setSymbolResolver(symbolSolver);
        CompilationUnit cu = StaticJavaParser.parse(FILE_TEXT);
        // Get the `stream.collect(...)` method call (the only method call in the example code).
        MethodCallExpr expr = cu.findAll(MethodCallExpr.class).get(0);

        // Running the following code will throw a RuntimeException with
        // `Error calculating the type of parameter Collectors.groupingBy(...)`
        // instead of returning the ResolvedType.
        ResolvedType resolvedType = expr.calculateResolvedType();

        Assertions.assertNotNull(resolvedType);
    }
}
```
Running the test will throw the RuntimeException. The complete StackTrace looks like this:
```
java.lang.RuntimeException: Error calculating the type of parameter Collectors.groupingBy(String::new, Collectors.counting()) of method call stream.collect(Collectors.groupingBy(String::new, Collectors.counting()))

	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:809)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:329)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:71)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:116)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:547)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:394)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:376)
	at com.github.javaparser.symbolsolver.JavaSymbolSolver.calculateType(JavaSymbolSolver.java:291)
	at com.github.javaparser.ast.expr.Expression.calculateResolvedType(Expression.java:550)
	at org.abego.playground.javaparserplayground.bug.parametertypenotcalculated.ParameterTypeNotCalculcatedTest.lambda$resolveType$0(ParameterTypeNotCalculcatedTest.java:35)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.abego.playground.javaparserplayground.bug.parametertypenotcalculated.ParameterTypeNotCalculcatedTest.resolveType(ParameterTypeNotCalculcatedTest.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:675)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:125)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:132)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:124)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:74)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:104)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:62)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:43)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:35)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:202)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:198)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:69)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:229)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:197)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:191)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.UnsupportedOperationException: com.github.javaparser.resolution.types.ResolvedWildcard
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.matchTypeParameters(MethodCallExprContext.java:412)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.matchTypeParameters(MethodCallExprContext.java:429)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.resolveMethodTypeParameters(MethodCallExprContext.java:389)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveMethodAsUsage(MethodCallExprContext.java:108)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:815)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:329)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:71)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:116)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:547)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:424)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:807)
	... 76 more
```

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/MethodCallExprContext.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/MethodCallExprContext.java
@@ -422,6 +422,32 @@ public class MethodCallExprContext extends ExpressionContext<MethodCallExpr> {
         return LeastUpperBoundLogic.of().lub(resolvedTypes);
     }
 
+    /**
+     * Attempts to match the formal type parameter {@code expectedType} against the actual argument
+     * type {@code actualType}, recording any resolved type-variable bindings in
+     * {@code matchedTypeParameters}.
+     *
+     * <p>Supported cases for {@code expectedType}:
+     * <ul>
+     *   <li><b>Type variable</b> – binds the actual type to the variable after boxing primitives and
+     *       replacing {@code null} with {@code Object}.
+     *       When {@code actualType} is a wildcard (e.g. the intermediate accumulation type {@code ?}
+     *       in {@code Collector<T, ?, R>}), capture-conversion rules apply: a bounded wildcard
+     *       ({@code ? extends Foo} or {@code ? super Foo}) contributes its declared bound as the
+     *       inferred type; an unbounded wildcard {@code ?} carries no type information and is
+     *       skipped.</li>
+     *   <li><b>Array</b> – recurses on the component type (null actual types pass through as-is,
+     *       see issue #2258).</li>
+     *   <li><b>Reference type</b> – recurses on each type argument when the actual type also
+     *       carries type arguments.</li>
+     *   <li><b>Primitive or wildcard</b> – no binding needed; returns immediately.</li>
+     * </ul>
+     *
+     * @param expectedType          formal parameter type, possibly containing type variables
+     * @param actualType            actual argument type at the call site
+     * @param matchedTypeParameters accumulator map from type-variable declarations to inferred types
+     * @throws UnsupportedOperationException if an unrecognised type combination is encountered
+     */
     private void matchTypeParameters(
             ResolvedType expectedType,
             ResolvedType actualType,
@@ -441,6 +467,16 @@ public class MethodCallExprContext extends ExpressionContext<MethodCallExpr> {
             if (type.isNull()) {
                 ResolvedReferenceTypeDeclaration resolvedTypedeclaration = typeSolver.getSolvedJavaLangObject();
                 type = new ReferenceTypeImpl(resolvedTypedeclaration);
+            }
+            // When the actual type is a wildcard (e.g. '?' in Collector<T, ?, R>), apply capture
+            // conversion: use the declared bound for bounded wildcards, and skip unbounded ones
+            // since no concrete type can be inferred from '?' alone (issue #3751).
+            if (type.isWildcard()) {
+                ResolvedWildcard wildcard = type.asWildcard();
+                if (wildcard.isBounded()) {
+                    matchedTypeParameters.put(expectedType.asTypeParameter(), wildcard.getBoundedType());
+                }
+                return;
             }
             if (!type.isTypeVariable() && !type.isReferenceType() && !type.isArray()) {
                 throw new UnsupportedOperationException(type.getClass().getCanonicalName());

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/javaparser/contexts/MethodCallExprContextResolutionTest.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/javaparser/contexts/MethodCallExprContextResolutionTest.java
@@ -203,6 +203,66 @@ class MethodCallExprContextResolutionTest extends AbstractResolutionTest {
         assertEquals("java.lang.String", resolvedType.describe());
     }
 
+    // Related to issue #3751
+    /**
+     * Verifies that {@code calculateResolvedType()} resolves the return type of
+     * {@code stream.collect(Collectors.toList())} without throwing an exception (issue #3751).
+     *
+     * <p>{@code Collectors.toList()} returns {@code Collector<T, ?, List<T>>}: the intermediate
+     * accumulation type is the unbounded wildcard {@code ?}.  {@code Stream.collect} is declared as
+     * {@code <R, A> R collect(Collector<? super T, A, R>)}, so resolving the return type requires
+     * matching the formal type variable {@code A} against the wildcard {@code ?}.
+     *
+     * <p>Before the fix, {@code matchTypeParameters} threw {@code UnsupportedOperationException}
+     * in this situation because it only accepted type variables, reference types, and arrays as
+     * candidates for type-variable bindings — wildcards were not handled.  The fix applies
+     * capture-conversion rules: bounded wildcards contribute their declared bound, and unbounded
+     * wildcards are skipped (no type information can be inferred from {@code ?} alone).  This
+     * leaves {@code A} unresolved while still allowing {@code R = List<String>} to be inferred
+     * correctly from the remaining type arguments.
+     */
+    @Test
+    void resolveStreamCollectWithWildcardAccumulatorType() {
+        ParserConfiguration config =
+                new ParserConfiguration().setSymbolResolver(new JavaSymbolSolver(new ReflectionTypeSolver()));
+        StaticJavaParser.setConfiguration(config);
+        CompilationUnit cu = parseSample("Issue3751");
+        List<MethodCallExpr> expressions = cu.getChildNodesByType(MethodCallExpr.class);
+
+        MethodCallExpr collectCall = expressions.stream()
+                .filter(e -> e.getNameAsString().equals("collect"))
+                .findFirst()
+                .orElseThrow(() -> new AssertionError("No 'collect' call found in Issue3751 sample"));
+
+        // Must not throw UnsupportedOperationException (issue #3751)
+        ResolvedType resolvedType = collectCall.calculateResolvedType();
+        assertEquals("java.util.List<java.lang.String>", resolvedType.describe());
+    }
+
+    /**
+     * Verifies that the wildcard fix does not regress the common case where all type arguments are
+     * concrete (no wildcards involved).  {@code Stream.collect(Collectors.toList())} returns
+     * {@code List<String>} which must still resolve correctly.
+     */
+    @Test
+    void resolveStreamCollectWithConcreteCollector() {
+        ParserConfiguration config =
+                new ParserConfiguration().setSymbolResolver(new JavaSymbolSolver(new ReflectionTypeSolver()));
+        StaticJavaParser.setConfiguration(config);
+
+        String code = "import java.util.*; import java.util.stream.*;"
+                + "class T { List<String> f(Stream<String> s) { return s.collect(Collectors.toList()); } }";
+        CompilationUnit cu = StaticJavaParser.parse(code);
+
+        MethodCallExpr collectCall = cu.getChildNodesByType(MethodCallExpr.class).stream()
+                .filter(e -> e.getNameAsString().equals("collect"))
+                .findFirst()
+                .orElseThrow(() -> new AssertionError("No 'collect' call found"));
+
+        ResolvedType resolvedType = collectCall.calculateResolvedType();
+        assertEquals("java.util.List<java.lang.String>", resolvedType.describe());
+    }
+
     // Related to issue #3195
     @Test
     void solveVariadicStaticGenericMethodCallCanInferFromArguments2() {

--- a/javaparser-symbol-solver-testing/src/test/resources/Issue3751.java.txt
+++ b/javaparser-symbol-solver-testing/src/test/resources/Issue3751.java.txt
@@ -1,0 +1,15 @@
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Issue3751 {
+
+    // Reproduces issue #3751: calculateResolvedType() on stream.collect() threw
+    // UnsupportedOperationException when the Collector argument contained a wildcard
+    // intermediate-accumulation type (the '?' in Collector<T, ?, R>).
+    // Collectors.toList() returns Collector<T, ?, List<T>>, where the accumulator '?'
+    // must be matched against the type variable 'A' in Stream.collect(Collector<? super T, A, R>).
+    List<String> toList(Stream<String> stream) {
+        return stream.collect(Collectors.toList());
+    }
+}

